core/trino-main/src/main/java/io/trino/sql/planner/BloomFilterWithRange.java

```diff
@@ -0,0 +1,303 @@
+            bloom = new long[bloomSize];
+            bloomSizeMask = bloomSize - 1;
+        }
+
+        private LongBloomFilter(long[] bloom, int minDistinctHashes)
+        {
+            this.bloom = bloom;
+            this.bloomSizeMask = bloom.length - 1;
+            this.minDistinctHashes = minDistinctHashes;
+        }
+
+        @JsonCreator
+        @DoNotCall // For JSON deserialization only
+        public static LongBloomFilter fromJson(
+                @JsonProperty("bloom") long[] bloom,
+                @JsonProperty("minDistinctHashes") int minDistinctHashes)
+        {
+            return new LongBloomFilter(bloom, minDistinctHashes);
+        }
+
+        @JsonProperty
+        public long[] getBloom()
+        {
+            return bloom;
+        }
+
+        @JsonProperty
+        public int getMinDistinctHashes()
+        {
+            return minDistinctHashes;
+        }
+
+        public boolean isEmpty()
+        {
+            return minDistinctHashes == 0;
+        }
+
+        public long getSizeInBytes()
+        {
+            return (long) bloom.length * Long.BYTES;
+        }
+
+        public void insert(long value)
+        {
+            long hashCode = XxHash64.hash(value);
+            long mask = bloomMask(hashCode);
+            int index = bloomIndex(hashCode);
+            if (mask == (bloom[index] & mask)) {
+                return;
+            }
+            // Set 3 bits in a 64 bit word
+            bloom[index] |= mask;
+            minDistinctHashes++;
+        }
+
+        public boolean contains(long value)
+        {
+            long hashCode = XxHash64.hash(value);
+            long mask = bloomMask(hashCode);
+            return mask == (bloom[bloomIndex(hashCode)] & mask);
+        }
+
+        public void merge(LongBloomFilter other)
+        {
+            checkArgument(bloom.length == other.bloom.length, "Bloom filters must have the same size");
+            minDistinctHashes += other.getMinDistinctHashes();
+            for (int i = 0; i < bloom.length; i++) {
+                bloom[i] |= other.bloom[i];
+            }
+        }
+
+        public void intersect(LongBloomFilter other)
+        {
+            checkArgument(bloom.length == other.bloom.length, "Bloom filters must have the same size");
+            minDistinctHashes = Math.min(minDistinctHashes, other.getMinDistinctHashes());
+            for (int i = 0; i < bloom.length; i++) {
+                bloom[i] &= other.bloom[i];
+            }
+        }
+
+        private int bloomIndex(long hashCode)
+        {
+            // Lower 21 bits are not used by bloomMask
+            // These are enough for the maximum size array that will be used here
+            return (int) (hashCode & bloomSizeMask);
+        }
+
+        private static long bloomMask(long hashCode)
+        {
+            // returned mask sets 3 bits based on portions of given hash
+            // Extract 38th to 43rd bits
```

Comment:
All this bit numbers is sheer magic. Please provide a structure of bloom filter in javadoc at the top of the class. 
Some ascii art with bit marking would be appreciated.